### PR TITLE
add methods for unstable LTIModels

### PR DESCRIPTION
**Pull request by lbalicki**
In this draft a few methods that enable model reduction for unstable LTI systems are added. Some of the main features include:
- System norms that support unstable systems (namely `l2_norm` and  `linf_norm`)
- The `FDBTReductor` using frequency domain representation of the gramians for balanced truncation
- `GapIRKAReductor`
- Bernoulli matrix equation solver and Bernoulli stabilization for LTI systems
- New options for `eigs`:
    - Shift and invert mode which allows computing eigenvalues near a specified (complex) shift
    - `complex_evp` assuming the input operator is complex
    - `left_evp` allows for computing left eigenvectors instead of right
- Dense Riccati equation solver via `solve_ricc_dense`

A few notes:
- Since there is multiple methods that require knowledge of the anti-stable system parts (i.e., system poles with positive real part) there is a caching mechanism through the new `_get_ast_spectrum` method and `_ast_spectrum` attribute of the `LTIModel` class
- The `hinf_norm` method calls the SLYCOT function `ab13dd`, which actually computes the L infinity norm (L infinity and H infinity are identical for asymptotically stable systems); So for now `linf_norm` just calls `hinf_norm`, maybe this should somehow be adjusted at some point 
- Support for Stokes-type systems will be in a separate PR

### CODE REVIEW

**Comment by sdrave** [comment]
>  Since there is multiple methods that require knowledge of the anti-stable system parts (i.e., system poles with positive real part) there is a caching mechanism through the new `_get_ast_spectrum` method and `_ast_spectrum` attribute of the `LTIModel` class

If I see correctly, your home-grown caching in `_get_ast_spectrum` seems to completely ignore the passed parameters. For `mu` this will clearly lead to mathematically wrong results. Also for `ast_pole_data` this does not look like a good idea. Is it really needed that you can pass the tuple version for `ast_pole_data`? Without this feature, you should be able to simply use pyMOR's caching.

**Comment by lbalicki** [comment]
With the latest commit pyMOR's caching is used for the anti-stable poles and the parameters (especially `mu`) are not ignored. 

The idea of the tuple option is that if a user uses a backend that has its own eigensolver (which computes poles and eigenvectors) they can convert the computed eigenvectors to `VectorArrays` and pass them as the argument to this function accordingly. Maybe there is a better option to enable this behaviour, especially since with the current approach users will need to create their own `VectorArrays` before passing them to the `_get_ast_spectrum` function.

**Comment by sdrave** [comment]
> The idea of the tuple option is that if a user uses a backend that has its own eigensolver (which computes poles and eigenvectors) they can convert the computed eigenvectors to `VectorArrays` and pass them as the argument to this function accordingly. Maybe there is a better option to enable this behaviour, especially since with the current approach users will need to create their own `VectorArrays` before passing them to the `_get_ast_spectrum` function.

The reason I'm bringing this up is that you can't call a `@cached` method with something that contains a `VectorArray` as argument.

**Comment by lbalicki** [comment]
> The reason I'm bringing this up is that you can't call a `@cached` method with something that contains a `VectorArray` as argument.

I will think about this more then. Since the caching of the anti-stable pole data is only needed in few cases, I might as well drop the `@cached` completely.

File: src/pymor/algorithms/bernoulli.py
Comment (pmli) [inline src/pymor/algorithms/bernoulli.py:41]:
```suggestion
        Whether to solve transposed or standard Bernoulli equation.
```

File: src/pymor/algorithms/bernoulli.py
Comment (pmli) [inline src/pymor/algorithms/bernoulli.py:78]: I'm worried that the determinant computation might overflow or underflow. Maybe use the logarithm? NumPy has [`slogdet`](https://numpy.org/doc/stable/reference/generated/numpy.linalg.slogdet.html#numpy.linalg.slogdet).

File: src/pymor/algorithms/bernoulli.py
Comment (pmli) [inline src/pymor/algorithms/bernoulli.py:96]: Delete?

File: src/pymor/algorithms/bernoulli.py
Comment (pmli) [inline src/pymor/algorithms/bernoulli.py:109]: Any reference for the method?

File: src/pymor/models/iosys.py
Comment (pmli) [inline src/pymor/models/iosys.py:939]: I'm not sure about using a mutable object (`dict`) as a default value (see [here](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments)).

File: src/pymor/reductors/bt.py
Comment (pmli) [inline src/pymor/reductors/bt.py:145]:
```suggestion
    """Balanced Truncation reductor using frequency domain representation of Gramians.
```

File: src/pymor/reductors/h2.py
Comment (pmli) [inline src/pymor/reductors/h2.py:685]: This depends on `conv_crit`, right?

**Review by pmli — COMMENTED** [review]
A couple of comments after a brief look.

**Comment by lbalicki** [comment]
I changed some of the points that @pmli mentioned and I also removed the caching of the anti-stable poles. I can only think of very few cases where the caching would be useful anyway and in case it is necessary the user can just store the eigenvectors in a VectorArray and pass them as an argument to the functions that require it.

File: src/pymor/reductors/h2.py
Comment (pmli) [inline src/pymor/reductors/h2.py:738]: It seems unfortunate to copy large parts of `GenericIRKAReductor`. Would it be possible to make `GenericIRKAReductor` even more generic and reduce copies here?

**Comment by pmli** [comment]
Besides the comment above, the code looks good to me. It would be nice to add a demo for the new MOR methods, once #1198 is merged. for testing purposes.

BTW, I was wondering if a dense solver for Bernoulli equations is available somewhere. It looks like SLICOT does not have it, which I found a bit surprising. On the other hand, C-M.E.S.S. does have it, but Py-M.E.S.S. does not wrap it. So, I guess it's fine to have a dense solver in pyMOR, although that's typically done.

**Comment by lbalicki** [comment]
> Besides the comment above, the code looks good to me. It would be nice to add a demo for the new MOR methods, once #1198 is merged. for testing purposes.

Do you have a suggestion of a simple unstable system that I can use in the demos?

> BTW, I was wondering if a dense solver for Bernoulli equations is available somewhere. It looks like SLICOT does not have it, which I found a bit surprising. On the other hand, C-M.E.S.S. does have it, but Py-M.E.S.S. does not wrap it. So, I guess it's fine to have a dense solver in pyMOR, although that's typically done.

I'm not sure about other Bernoulli equation solvers out there, but a Bernoulli equation is actually just a special case of a Riccati equation. In that sense it might be worth testing the performance of dense Riccati equation solvers against the Bernoulli solver I implemented for the PR.

**Comment by pmli** [comment]
> Do you have a suggestion of a simple unstable system that I can use in the demos?

Not really... Maybe one of the cited papers has a simple enough model? Otherwise, you could construct a synthetic example similar to Penzl's FOM example.

> I'm not sure about other Bernoulli equation solvers out there, but a Bernoulli equation is actually just a special case of a Riccati equation. In that sense it might be worth testing the performance of dense Riccati equation solvers against the Bernoulli solver I implemented for the PR.

Right, I was actually wondering if Riccati equation solvers break down for the Bernoulli equation. It could be worth trying.

**Comment by drittelhacker** [comment]
> Right, I was actually wondering if Riccati equation solvers break down for the Bernoulli equation. It could be worth trying.
careful with that. Newton based solvers need an initial stabilizing feedback if the LTI is not stable. That is where often Bernoulli comes into play, so for Newton your idea is biting it's own tail. 

Solvers based on the Hamiltonian EVP should do better, as long as you can make sure the eigenvalues stay away from the imaginary axis.

File: src/pymor/reductors/h2.py
Comment (lbalicki) [reply src/pymor/reductors/h2.py:738]: I could give the `_rom_to_sigma_b_c` method of `GenericIRKAReductor` an extra output ( i.e. change `return sigma, b, c` to `return sigma, b, c, None`). The `None` output can then be changed to some relevant value in specialized classes, e.g. in `GapIRKAReductor` the `None` would become `gap_rom`. This way I could eventually completely remove the `_rom0_params_to_sigma_b_c` method from `GapIRKAReductor`, is this what you mean by more generic?

File: src/pymor/reductors/h2.py
Comment (pmli) [inline src/pymor/reductors/h2.py:786]: It would be good to add support for other solvers. This looks to me like good motivation to add `algorithms.riccati.solve_ricc_dense` (similar to `algorithms.lyapunov.solve_lyap_dense`).

File: src/pymor/reductors/h2.py
Comment (pmli) [inline src/pymor/reductors/h2.py:754]: Maybe I'm missing something, but is there a reason for not using `LTIBHIReductor`?

File: src/pymor/reductors/h2.py
Comment (pmli) [inline src/pymor/reductors/h2.py:707]: It looks like this is copied from `OneSidedIRKAReductor`, but it should be more similar to `IRKAReductor` since there are two projection matrices (in particular, `'biorth'` should be used instead of `'Eorth'`).

File: src/pymor/reductors/h2.py
Comment (pmli) [reply src/pymor/reductors/h2.py:738]: Hmm, now that I looked more closely at other methods you needed to change, maybe it would be better to do the refactoring in another PR and focus here on just adding the methods. I'm not completely happy with the state of H2-type methods and this new method is good motivation to re-think the structure.

**Comment by lbalicki** [comment]
I adressed some of the changes @pmli mentioned. Also, I added dense Riccati equation solvers (for now only scipy and slycot are supported) and there is a short tutorial on how to use the new methods.

File: src/pymor/algorithms/eigs.py
Comment (pmli) [inline src/pymor/algorithms/eigs.py:201]:
```suggestion
    H = np.zeros((l, l), dtype=np.complex_ if complex_evp else np.float_)
```

File: src/pymor/bindings/scipy.py
Comment (pmli) [inline src/pymor/bindings/scipy.py:512]: "Low-rank" and "lrcf" should be replaced.

File: docs/source/tutorial_unstable_lti_systems.rst
Comment (pmli) [inline docs/source/tutorial_unstable_lti_systems.rst:124]:
```suggestion
        \operatorname{d}\!\omega E.
```

**Review by pmli — COMMENTED** [review]
It's looking good. Below are some minor comments. Here are two more things:

- `solve_ricc_lrcf` in `scipy` and `slycot` bindings could use `solve_ricc_dense` to reduce code duplication,
- it would be good to add a demo (e.g., `unstable_heat` or similar) to include the new reductors in the coverage report.

File: src/pymor/bindings/slycot.py
Comment (pmli) [inline src/pymor/bindings/slycot.py:241]: It seems this function is unused.

File: src/pymor/algorithms/bernoulli.py
Comment (pmli) [inline src/pymor/algorithms/bernoulli.py:54]: There should be no empty line after a function docstring, but the PR linter is not reporting it. If #1278 works, rebasing on main after that's merged should help.

**Review by pmli — COMMENTED** [review]
Just a few more comments.

File: src/pymor/models/iosys.py
Comment (pmli) [inline src/pymor/models/iosys.py:1058]:
```suggestion
            ew, lev, rev = spla.eig(A, E if self.E else None, left=True)
```

File: src/pymor/models/iosys.py
Comment (pmli) [inline src/pymor/models/iosys.py:1060]:
```suggestion
            ast_ews = ew[ew.real > 0]
```

**Review by pmli — COMMENTED** [review]
Some comments on `get_ast_spectrum`. That should be it...

File: src/pymor/models/iosys.py
Comment (lbalicki) [reply src/pymor/models/iosys.py:1025]: Maybe I am missing something but I need `ast_idx` as an index later in the function, so I left it as is.

Comment (pmli) [reply src/pymor/models/iosys.py:1025]: Aha, right, ok then.

File: src/pymor/models/iosys.py
Comment (pmli) [inline src/pymor/models/iosys.py:1027]: For a consistent return type:

```suggestion
                    return (self.solution_space.empty(), np.empty((0,)), self.solution_space.empty())
```

**Review by pmli — COMMENTED** [review]
One more thing that I was reminded of, is running the demo during testing. It should be enough to extend `SYS_MOR_ARGS` in `src/pymortests/demos.py`.

**Comment by pmli** [comment]
@lbalicki Can this draft PR be turned into a regular PR?
@sdrave Do you see anything that should be worked on?

File: src/pymor/models/iosys.py
Comment (sdrave) [inline src/pymor/models/iosys.py:923]: Maybe add an definition of the norm?

File: src/pymor/models/iosys.py
Comment (sdrave) [inline src/pymor/models/iosys.py:974]: Maybe add an definition of the norm?

Comment (sdrave) [reply src/pymor/models/iosys.py:974]: And explain why this is the same as calling `hinf_norm`?

File: src/pymor/reductors/bt.py
Comment (sdrave) [inline src/pymor/reductors/bt.py:190]: Add some documentation specifying which error is actually bounded?

File: src/pymor/algorithms/bernoulli.py
Comment (lbalicki) [reply src/pymor/algorithms/bernoulli.py:92]: I am not sure if I would be for failing all iterative algorithms if they don't reach the tolerance within the maximum iteration number. I have some examples in mind where giving a warning would be more desirable. So we should probably discuss this more thoroughly at some point.

File: src/pymor/algorithms/riccati.py
Comment (lbalicki) [reply src/pymor/algorithms/riccati.py:172]: This should probably be changed for all the dense Lyapunov solvers as well then.

File: src/pymor/models/iosys.py
Comment (lbalicki) [reply src/pymor/models/iosys.py:974]: Currently `LTIModel.hinf_norm` is based on slycot's [`ab13dd`](http://slicot.org/objects/software/shared/doc/AB13DD.html) which computes the L-infinity norm (there is no slycot function for H-infinity norm computation). For asymptotically stable systems H-inf and L-inf norms are identical, which is why we use it that way. It will probably make sense to put the current `hinf_norm` code into `linf_norm` method and have `hinf_norm` call  `linf_norm` rather than the other way around. Should we test for asymptotic stability in `hinf_norm` somehow such that it is more accurate?

Comment (lbalicki) [reply src/pymor/models/iosys.py]: @pmli Do you have an opinion on this?

Comment (pmli) [reply src/pymor/models/iosys.py]: I'm not sure, but since `hinf_norm` works only for dense systems, I guess it doesn't hurt too much to add a stability check using `eig`, and return `np.inf` if the system is not asymptotically stable.

File: src/pymor/algorithms/bernoulli.py
Comment (pmli) [reply src/pymor/algorithms/bernoulli.py:94]: I agree that it would be good to discuss this. At least, this is consistent with low-rank Lyapunov and Riccati solvers and IRKA.

File: src/pymor/models/iosys.py
Comment (lbalicki) [reply src/pymor/models/iosys.py]: After doing some testing, it turned out that `eig` is very slow. E.g., for a 1000 x 1000 matrix `eig` takes roughly 60 seconds whereas in comparison `linf_norm` takes 0.0005 seconds. Since we have the note 'Assumes the system is asymptotically stable.' in the `hinf_norm` documentation, I would be for not adding the stability check for now.

File: src/pymor/algorithms/bernoulli.py
Comment (sdrave) [reply src/pymor/algorithms/bernoulli.py:94]: I have opened an issue: #1372.

File: src/pymor/models/iosys.py
Comment (sdrave) [reply src/pymor/models/iosys.py]: > It will probably make sense to put the current `hinf_norm` code into `linf_norm` method and have `hinf_norm` call `linf_norm` rather than the other way around. 

Agreed. Can you do that in this PR?

> Since we have the note 'Assumes the system is asymptotically stable.' in the hinf_norm documentation, I would be for not adding the stability check for now.

I'm fine with this. Maybe extend the documentation a little and say that under this assumption both norms are the same, so that we just defer to the other method?

**Review by sdrave — APPROVED** [review]

**Pull request merged**
